examples/web_api.rs

```diff
@@ -10,80 +10,87 @@ use hyper::{Body, Chunk, Client, Method, Request, Response, Server, StatusCode,
 use hyper::client::HttpConnector;
 use hyper::service::service_fn;
 
-#[allow(unused, deprecated)]
-use std::ascii::AsciiExt;
-
 static NOTFOUND: &[u8] = b"Not Found";
-static URL: &str = "http://127.0.0.1:1337/web_api";
+static URL: &str = "http://127.0.0.1:1337/json_api";
 static INDEX: &[u8] = b"<a href=\"test.html\">test.html</a>";
-static LOWERCASE: &[u8] = b"i am a lower case string";
+static POST_DATA: &str = r#"{"original": "data"}"#;
+
+type ResponseFuture = Box<Future<Item=Response<Body>, Error=hyper::Error> + Send>;
+
+fn client_request_response(client: &Client<HttpConnector>) -> ResponseFuture {
+     let req = Request::builder()
+         .method(Method::POST)
+         .uri(URL)
+         .header(header::CONTENT_TYPE, "application/json")
+         .body(POST_DATA.into())
+         .unwrap();
+
+     Box::new(client.request(req).map(|web_res| {
+         // Compare the JSON we sent (before) with what we received (after):
+         let body = Body::wrap_stream(web_res.into_body().map(|b| {
+             Chunk::from(format!("<b>POST request body</b>: {}<br><b>Response</b>: {}",
+                                 POST_DATA,
+                                 std::str::from_utf8(&b).unwrap()))
+         }));
+
+         Response::new(body)
+     }))
+}
+
+fn api_post_response(req: Request<Body>) -> ResponseFuture {
+    // A web api to run against
+    Box::new(req.into_body()
+        .concat2() // Concatenate all chunks in the body
+        .and_then(|entire_body| {
+            // TODO: Replace all unwraps with proper error handling
```

Comment:
If the error type of `ResponseFuture` were changed to `Box<dyn std::error::Error + Send + Sync>`, then these `unwrap`s can be replaced with `?`. And the final `future::ok(response)` can just be `Ok(response)`.

Reply:
```
type GenericError = Box<dyn std::error::Error + Send + Sync>;
type ResponseFuture = Box<Future<Item=Response<Body>, Error=GenericError> + Send>;

fn client_request_response(client: &Client<HttpConnector>) -> ResponseFuture {
     let req = Request::builder()
         .method(Method::POST)
         .uri(URL)
         .header(header::CONTENT_TYPE, "application/json")
         .body(POST_DATA.into())
         .unwrap();

     Box::new(client.request(req).map(|web_res| {
         // Compare the JSON we sent (before) with what we received (after):
         let body = Body::wrap_stream(web_res.into_body().map(|b| {
             Chunk::from(format!("<b>POST request body</b>: {}<br><b>Response</b>: {}",
                                 POST_DATA,
                                 std::str::from_utf8(&b).unwrap()))
         }));

         Response::new(body)
     }))
}
```

```
olehermanse@OHMBP-2869 hyper $ cargo run --example web_api
   Compiling hyper v0.12.25 (/Users/olehermanse/code/hyper)
error[E0271]: type mismatch resolving `<futures::Map<hyper::client::ResponseFuture, [closure@examples/web_api.rs:29:39: 38:7]> as futures::Future>::Error == std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>`
  --> examples/web_api.rs:29:6
   |
29 | /      Box::new(client.request(req).map(|web_res| {
30 | |          // Compare the JSON we sent (before) with what we received (after):
31 | |          let body = Body::wrap_stream(web_res.into_body().map(|b| {
32 | |              Chunk::from(format!("<b>POST request body</b>: {}<br><b>Response</b>: {}",
...  |
37 | |          Response::new(body)
38 | |      }))
   | |________^ expected struct `hyper::Error`, found struct `std::boxed::Box`
   |
   = note: expected type `hyper::Error`
              found type `std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>`
   = note: required for the cast to the object type `dyn futures::Future<Item=hyper::Response<hyper::Body>, Error=std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>> + std::marker::Send`

error[E0271]: type mismatch resolving `<futures::AndThen<futures::stream::Concat2<hyper::Body>, std::result::Result<hyper::Response<hyper::Body>, hyper::Error>, [closure@examples/web_api.rs:45:19: 57:10]> as futures::Future>::Error == std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>`
  --> examples/web_api.rs:43:5
   |
43 | /     Box::new(req.into_body()
44 | |         .concat2() // Concatenate all chunks in the body
45 | |         .and_then(|entire_body| {
46 | |             // TODO: Replace all unwraps with proper error handling
...  |
57 | |         })
58 | |     )
   | |_____^ expected struct `hyper::Error`, found struct `std::boxed::Box`
   |
   = note: expected type `hyper::Error`
              found type `std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>`
   = note: required for the cast to the object type `dyn futures::Future<Item=hyper::Response<hyper::Body>, Error=std::boxed::Box<dyn std::error::Error + std::marker::Send + std::marker::Sync>> + std::marker::Send`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0271`.
error: Could not compile `hyper`.

To learn more, run the command again with --verbose.
olehermanse@OHMBP-2869 hyper $ 
```

Reply:
@seanmonstar That change causes other problems that I have no clue how to fix ^

Reply:
Oh sure. That's because `and_then` wants the error types to be the same, and the concat2 error is `hyper::Error`.

You can put a `.concat2().map_err(From::from)` so that the `and_then` can return a different error type. I know, it sounds like magic...

Reply:
@seanmonstar Alright, thanks. I also found `.from_err()` which seems to do the same.

Then, `?` operator doesn't really work for futures:

```
olehermanse@OHMBP-2869 hyper $ cargo run --example web_api
   Compiling hyper v0.12.25 (/Users/olehermanse/code/hyper)
error[E0277]: the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
  --> examples/web_api.rs:48:23
   |
48 |             let str = String::from_utf8(entire_body.to_vec())?;
   |                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `futures::FutureResult<hyper::Response<hyper::Body>, _>`
   |
   = help: the trait `std::ops::Try` is not implemented for `futures::FutureResult<hyper::Response<hyper::Body>, _>`
   = note: required by `std::ops::Try::from_error`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `hyper`.

To learn more, run the command again with --verbose.
olehermanse@OHMBP-2869 hyper $ 
```

Reply:
You gotta change from `future::ok(response)` to just `Ok(response)`. The `?` will early-return `Err(err)` if it's an error, which is a `Result`. So anything else that is returned from the same closure must also be a `Result`, like `Ok(response)`.

Reply:
Thanks, fixed.